File: claw-buddy-backend/app/schemas/gene.py
"""Pydantic schemas for Gene Evolution Ecosystem."""

from datetime import datetime

from pydantic import BaseModel, Field


# ── Gene ─────────────────────────────────────────


class GeneInfo(BaseModel):
    id: str
    name: str
    slug: str
    description: str | None = None
    short_description: str | None = None
    category: str | None = None
    tags: list[str] = []
    source: str = "official"
    source_ref: str | None = None
    icon: str | None = None
    version: str = "1.0.0"
    manifest: dict | None = None
    dependencies: list[str] = []
    synergies: list[str] = []
    parent_gene_id: str | None = None
    created_by_instance_id: str | None = None
    install_count: int = 0
    avg_rating: float = 0.0
    effectiveness_score: float = 0.0
    is_featured: bool = False
    review_status: str | None = None
    is_published: bool = True
    created_by: str | None = None
    org_id: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class GeneListItem(BaseModel):
    id: str
    name: str
    slug: str
    short_description: str | None = None
    category: str | None = None
    tags: list[str] = []
    source: str = "official"
    icon: str | None = None
    version: str = "1.0.0"
    install_count: int = 0
    avg_rating: float = 0.0
    effectiveness_score: float = 0.0
    is_featured: bool = False
    parent_gene_id: str | None = None
    created_by_instance_id: str | None = None

    model_config = {"from_attributes": True}


class GeneCreateRequest(BaseModel):
    name: str = Field(..., max_length=128)
    slug: str = Field(..., max_length=128)
    description: str | None = None
    short_description: str | None = Field(
        None, max_length=256,
        description="当 manifest 包含 skill 时必填，用于 OpenClaw SKILL.md YAML front matter",
    )
    category: str | None = Field(None, max_length=32)
    tags: list[str] = []
    source: str = "official"
    source_ref: str | None = None
    icon: str | None = Field(None, max_length=32)
    version: str = Field("1.0.0", max_length=16)
    manifest: dict | None = None
    dependencies: list[str] = []
    synergies: list[str] = []
    is_featured: bool = False
    is_published: bool = True


# ── Genome ───────────────────────────────────────


class GenomeInfo(BaseModel):
    id: str
    name: str
    slug: str
    description: str | None = None
    short_description: str | None = None
    icon: str | None = None
    gene_slugs: list[str] = []
    config_override: dict | None = None
    install_count: int = 0
    avg_rating: float = 0.0
    is_featured: bool = False
    is_published: bool = True
    created_by: str | None = None
    org_id: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class GenomeListItem(BaseModel):
    id: str
    name: str
    slug: str
    short_description: str | None = None
    icon: str | None = None
    gene_slugs: list[str] = []
    install_count: int = 0
    avg_rating: float = 0.0
    is_featured: bool = False

    model_config = {"from_attributes": True}


class GenomeCreateRequest(BaseModel):
    name: str = Field(..., max_length=128)
    slug: str = Field(..., max_length=128)
    description: str | None = None
    short_description: str | None = Field(None, max_length=256)
    icon: str | None = Field(None, max_length=32)
    gene_slugs: list[str] = []
    config_override: dict | None = None
    is_featured: bool = False
    is_published: bool = True


# ── Admin Update Requests ────────────────────────


class UpdateGeneRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    short_description: str | None = None
    category: str | None = None
    tags: list[str] | None = None
    icon: str | None = None
    version: str | None = None
    manifest: dict | None = None
    is_featured: bool | None = None
    is_published: bool | None = None


class UpdateGenomeRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    short_description: str | None = None
    icon: str | None = None
    gene_slugs: list[str] | None = None
    config_override: dict | None = None
    is_featured: bool | None = None
    is_published: bool | None = None


# ── InstanceGene ─────────────────────────────────


class InstanceGeneInfo(BaseModel):
    id: str
    instance_id: str
    gene_id: str
    genome_id: str | None = None
    status: str = "installing"
    installed_version: str | None = None
    learning_output: str | None = None
    config_snapshot: dict | None = None
    agent_self_eval: float | None = None
    usage_count: int = 0
    variant_published: bool = False
    installed_at: datetime | None = None
    created_at: datetime | None = None

    gene: GeneListItem | None = None

    model_config = {"from_attributes": True}


class InstallGeneRequest(BaseModel):
    gene_slug: str


class ApplyGenomeRequest(BaseModel):
    genome_id: str


class UninstallGeneRequest(BaseModel):
    gene_id: str


# ── Rating ───────────────────────────────────────


class RatingRequest(BaseModel):
    rating: int = Field(..., ge=1, le=5)
    comment: str | None = None


class RatingInfo(BaseModel):
    id: str
    user_id: str
    rating: int
    comment: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


# ── Effect ───────────────────────────────────────


class EffectivenessRequest(BaseModel):
    metric_type: str
    value: float = 1.0
    context: str | None = None


class GeneEffectLogInfo(BaseModel):
    id: str
    instance_id: str
    gene_id: str
    metric_type: str
    value: float
    context: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


# ── Evolution ────────────────────────────────────


class PublishVariantRequest(BaseModel):
    variant_name: str | None = None
    variant_slug: str | None = None


class CreateGeneRequest(BaseModel):
    creation_prompt: str | None = None


# ── Learning Callback ────────────────────────────


class LearningCallbackPayload(BaseModel):
    task_id: str
    instance_id: str
    mode: str  # "learn" | "create" | "forget"
    decision: str  # "direct_install" | "learned" | "failed" | "created" | "forgotten" | "simplified" | "forget_failed"
    content: str | None = None
    self_eval: float | None = None
    meta: dict | None = None
    reason: str | None = None


# ── Evolution Log ────────────────────────────────


class EvolutionEventInfo(BaseModel):
    id: str
    instance_id: str
    event_type: str
    gene_name: str
    gene_slug: str | None = None
    gene_id: str | None = None
    genome_id: str | None = None
    details: dict | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


# ── Admin ────────────────────────────────────────


class GeneStatsResponse(BaseModel):
    total_genes: int = 0
    total_installs: int = 0
    learning_count: int = 0
    failed_count: int = 0
    agent_created_count: int = 0


class ReviewRequest(BaseModel):
    action: str  # "approve" | "reject"
    reason: str | None = None


class TagStats(BaseModel):
    tag: str
    count: int


class CoInstallPair(BaseModel):
    gene_a_slug: str
    gene_b_slug: str
    co_install_count: int

File: claw-buddy-frontend/src/stores/gene.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/services/api'

export interface GeneItem {
  id: string
  name: string
  slug: string
  description: string | null
  short_description: string | null
  category: string | null
  tags: string[]
  source: string
  source_ref: string | null
  icon: string | null
  version: string
  manifest: Record<string, unknown> | null
  dependencies: string[]
  synergies: string[]
  parent_gene_id: string | null
  created_by_instance_id: string | null
  install_count: number
  avg_rating: number
  effectiveness_score: number
  is_featured: boolean
  review_status: string | null
  is_published: boolean
  created_by: string | null
  org_id: string | null
  created_at: string | null
  updated_at: string | null
}

export interface GenomeItem {
  id: string
  name: string
  slug: string
  description: string | null
  short_description: string | null
  icon: string | null
  gene_slugs: string[]
  config_override: Record<string, unknown> | null
  install_count: number
  avg_rating: number
  is_featured: boolean
  is_published: boolean
  created_by: string | null
  org_id: string | null
  created_at: string | null
}

export interface GeneStats {
  total_genes: number
  total_installs: number
  learning_count: number
  failed_count: number
  agent_created_count: number
}

export interface ActivityItem {
  id: string
  gene_slug: string
  gene_name: string
  metric_type: string
  value: number
  created_at: string | null
}

export const useGeneStore = defineStore('gene', () => {
  const genes = ref<GeneItem[]>([])
  const genomes = ref<GenomeItem[]>([])
  const totalGenes = ref(0)
  const totalGenomes = ref(0)
  const stats = ref<GeneStats | null>(null)
  const activity = ref<ActivityItem[]>([])
  const pendingGenes = ref<GeneItem[]>([])
  const loading = ref(false)

  async function fetchGenes(params: {
    keyword?: string
    category?: string
    is_published?: boolean
    sort?: string
    page?: number
    page_size?: number
  } = {}) {
    loading.value = true
    try {
      const res = await api.get('/admin/genes', { params })
      genes.value = res.data.data ?? []
      totalGenes.value = res.data.pagination?.total ?? 0
    } finally {
      loading.value = false
    }
  }

  async function fetchGenomes(params: {
    keyword?: string
    is_published?: boolean
    page?: number
    page_size?: number
  } = {}) {
    loading.value = true
    try {
      const res = await api.get('/admin/genomes', { params })
      genomes.value = res.data.data ?? []
      totalGenomes.value = res.data.pagination?.total ?? 0
    } finally {
      loading.value = false
    }
  }

  async function fetchStats() {
    const res = await api.get('/admin/genes/stats')
    stats.value = res.data.data
  }

  async function fetchActivity(limit = 50) {
    const res = await api.get('/admin/genes/activity', { params: { limit } })
    activity.value = res.data.data ?? []
  }

  async function fetchPendingGenes() {
    const res = await api.get('/admin/genes/pending-review')
    pendingGenes.value = res.data.data ?? []
  }

  async function createGene(data: Record<string, unknown>) {
    const res = await api.post('/admin/genes', data)
    return res.data.data
  }

  async function updateGene(geneId: string, data: Record<string, unknown>) {
    const res = await api.put(`/admin/genes/${geneId}`, data)
    return res.data.data
  }

  async function deleteGene(geneId: string) {
    await api.delete(`/admin/genes/${geneId}`)
  }

  async function reviewGene(geneId: string, action: string, reason?: string) {
    const res = await api.put(`/admin/genes/${geneId}/review`, { action, reason })
    return res.data.data
  }

  async function createGenome(data: Record<string, unknown>) {
    const res = await api.post('/admin/genomes', data)
    return res.data.data
  }

  async function updateGenome(genomeId: string, data: Record<string, unknown>) {
    const res = await api.put(`/admin/genomes/${genomeId}`, data)
    return res.data.data
  }

  async function deleteGenome(genomeId: string) {
    await api.delete(`/admin/genomes/${genomeId}`)
  }

  return {
    genes,
    genomes,
    totalGenes,
    totalGenomes,
    stats,
    activity,
    pendingGenes,
    loading,
    fetchGenes,
    fetchGenomes,
    fetchStats,
    fetchActivity,
    fetchPendingGenes,
    createGene,
    updateGene,
    deleteGene,
    reviewGene,
    createGenome,
    updateGenome,
    deleteGenome,
  }
})
